File: gnitz/dbsp/expr.py
# gnitz/dbsp/expr.py
#
# Expression bytecode definitions and Rust FFI wrappers for SQL scalar functions.

from rpython.rlib.rarithmetic import r_int64, intmask
from rpython.rlib.longlong2float import float2longlong
from rpython.rlib.objectmodel import newlist_hint

# ---------------------------------------------------------------------------
# Expression Opcodes (separate namespace from DBSP VM opcodes)
# ---------------------------------------------------------------------------

# Column loads
EXPR_LOAD_COL_INT   = 1   # dst = accessor.get_int_signed(arg1)
EXPR_LOAD_COL_FLOAT = 2   # dst = float2longlong(accessor.get_float(arg1))
EXPR_LOAD_CONST     = 3   # dst = arg1 (immediate i64 value)

# Integer arithmetic
EXPR_INT_ADD = 4
EXPR_INT_SUB = 5
EXPR_INT_MUL = 6
EXPR_INT_DIV = 7
EXPR_INT_MOD = 8
EXPR_INT_NEG = 9

# Float arithmetic (registers hold float bits via float2longlong)
EXPR_FLOAT_ADD = 10
EXPR_FLOAT_SUB = 11
EXPR_FLOAT_MUL = 12
EXPR_FLOAT_DIV = 13
EXPR_FLOAT_NEG = 14

# Integer comparison (signed i64)
EXPR_CMP_EQ = 15
EXPR_CMP_NE = 16
EXPR_CMP_GT = 17
EXPR_CMP_GE = 18
EXPR_CMP_LT = 19
EXPR_CMP_LE = 20

# Float comparison (interprets register bits as f64)
EXPR_FCMP_EQ = 21
EXPR_FCMP_NE = 22
EXPR_FCMP_GT = 23
EXPR_FCMP_GE = 24
EXPR_FCMP_LT = 25
EXPR_FCMP_LE = 26

# Boolean logic
EXPR_BOOL_AND = 27
EXPR_BOOL_OR  = 28
EXPR_BOOL_NOT = 29

# Null checks
EXPR_IS_NULL     = 30   # dst = accessor.is_null(arg1) ? 1 : 0
EXPR_IS_NOT_NULL = 31

# Output opcodes (Phase 4: computed projections)
EXPR_EMIT          = 32   # builder.append_int(regs[a1]) or append_null(a2)
EXPR_INT_TO_FLOAT  = 33   # regs[dst] = float2longlong(float(intmask(regs[a1])))
EXPR_COPY_COL      = 34   # type-dispatched direct copy: input col -> output col

# Fused string comparison opcodes (Phase 5: string predicates)
EXPR_STR_COL_EQ_CONST = 40   # dst = (col[a1] == const_strings[a2]) ? 1 : 0
EXPR_STR_COL_LT_CONST = 41   # dst = (col[a1] <  const_strings[a2]) ? 1 : 0
EXPR_STR_COL_LE_CONST = 42   # dst = (col[a1] <= const_strings[a2]) ? 1 : 0
EXPR_STR_COL_EQ_COL   = 43   # dst = (col[a1] == col[a2]) ? 1 : 0
EXPR_STR_COL_LT_COL   = 44   # dst = (col[a1] <  col[a2]) ? 1 : 0
EXPR_STR_COL_LE_COL   = 45   # dst = (col[a1] <= col[a2]) ? 1 : 0

# Null emission (Phase 6: LEFT JOIN null-fill)
EXPR_EMIT_NULL = 46   # builder.append_null(a1)  -- a1 = payload_col_idx


# ---------------------------------------------------------------------------
# ExprProgram -- immutable bytecode for JIT
# ---------------------------------------------------------------------------

class ExprProgram(object):
    _immutable_fields_ = ['code[*]', 'num_regs', 'result_reg', 'num_instrs',
                          'const_strings[*]', 'const_prefixes[*]',
                          'const_lengths[*]']

    def __init__(self, code, num_regs, result_reg, const_strings=None):
        from gnitz.core import strings as _strings
        self.code = code[:]         # copy: [*] requires a never-resized list
        self.num_regs = num_regs
        self.result_reg = result_reg
        self.num_instrs = len(code) // 4
        if const_strings is not None:
            self.const_strings = const_strings[:]
            prefixes = [0] * len(const_strings)
            lengths = [0] * len(const_strings)
            for _i in range(len(const_strings)):
                prefixes[_i] = _strings.compute_prefix(const_strings[_i])
                lengths[_i] = len(const_strings[_i])
            self.const_prefixes = prefixes
            self.const_lengths = lengths
        else:
            # RPython annotation: must be same type as the 'not None' branch
            self.const_strings = [""][0:0]
            self.const_prefixes = [0][0:0]
            self.const_lengths = [0][0:0]

    def code_as_ints(self):
        result = []
        for w in self.code:
            result.append(intmask(w))
        return result


# ---------------------------------------------------------------------------
# ExprBuilder -- ergonomic API for constructing expression programs
# ---------------------------------------------------------------------------

class ExprBuilder(object):
    def __init__(self):
        self._code = newlist_hint(64)
        self._next_reg = 0

    def _alloc_reg(self):
        r = self._next_reg
        self._next_reg += 1
        return r

    def _emit(self, op, dst, a1, a2):
        self._code.append(r_int64(op))
        self._code.append(r_int64(dst))
        self._code.append(r_int64(a1))
        self._code.append(r_int64(a2))

    # --- Column loads ---

    def load_col_int(self, col_idx):
        dst = self._alloc_reg()
        self._emit(EXPR_LOAD_COL_INT, dst, col_idx, 0)
        return dst

    def load_col_float(self, col_idx):
        dst = self._alloc_reg()
        self._emit(EXPR_LOAD_COL_FLOAT, dst, col_idx, 0)
        return dst

    def load_const_int(self, value):
        dst = self._alloc_reg()
        v = r_int64(value)
        lo = intmask(v)
        hi = intmask(v >> 32)
        self._emit(EXPR_LOAD_CONST, dst, lo, hi)
        return dst

    def load_const_float(self, value):
        dst = self._alloc_reg()
        bits = float2longlong(value)
        lo = intmask(bits)
        hi = intmask(bits >> 32)
        self._emit(EXPR_LOAD_CONST, dst, lo, hi)
        return dst

    # --- Integer arithmetic ---

    def int_add(self, src1, src2):
        dst = self._alloc_reg()
        self._emit(EXPR_INT_ADD, dst, src1, src2)
        return dst

    def int_sub(self, src1, src2):
        dst = self._alloc_reg()
        self._emit(EXPR_INT_SUB, dst, src1, src2)
        return dst

    def int_mul(self, src1, src2):
        dst = self._alloc_reg()
        self._emit(EXPR_INT_MUL, dst, src1, src2)
        return dst

    def int_div(self, src1, src2):
        dst = self._alloc_reg()
        self._emit(EXPR_INT_DIV, dst, src1, src2)
        return dst

    def int_mod(self, src1, src2):
        dst = self._alloc_reg()
        self._emit(EXPR_INT_MOD, dst, src1, src2)
        return dst

    def int_neg(self, src):
        dst = self._alloc_reg()
        self._emit(EXPR_INT_NEG, dst, src, 0)
        return dst

    # --- Float arithmetic ---

    def float_add(self, src1, src2):
        dst = self._alloc_reg()
        self._emit(EXPR_FLOAT_ADD, dst, src1, src2)
        return dst

    def float_sub(self, src1, src2):
        dst = self._alloc_reg()
        self._emit(EXPR_FLOAT_SUB, dst, src1, src2)
        return dst

    def float_mul(self, src1, src2):
        dst = self._alloc_reg()
        self._emit(EXPR_FLOAT_MUL, dst, src1, src2)
        return dst

    def float_div(self, src1, src2):
        dst = self._alloc_reg()
        self._emit(EXPR_FLOAT_DIV, dst, src1, src2)
        return dst

    def float_neg(self, src):
        dst = self._alloc_reg()
        self._emit(EXPR_FLOAT_NEG, dst, src, 0)
        return dst

    # --- Integer comparison ---

    def cmp_eq(self, src1, src2):
        dst = self._alloc_reg()
        self._emit(EXPR_CMP_EQ, dst, src1, src2)
        return dst

    def cmp_ne(self, src1, src2):
        dst = self._alloc_reg()
        self._emit(EXPR_CMP_NE, dst, src1, src2)
        return dst

    def cmp_gt(self, src1, src2):
        dst = self._alloc_reg()
        self._emit(EXPR_CMP_GT, dst, src1, src2)
        return dst

    def cmp_ge(self, src1, src2):
        dst = self._alloc_reg()
        self._emit(EXPR_CMP_GE, dst, src1, src2)
        return dst

    def cmp_lt(self, src1, src2):
        dst = self._alloc_reg()
        self._emit(EXPR_CMP_LT, dst, src1, src2)
        return dst

    def cmp_le(self, src1, src2):
        dst = self._alloc_reg()
        self._emit(EXPR_CMP_LE, dst, src1, src2)
        return dst

    # --- Float comparison ---

    def fcmp_eq(self, src1, src2):
        dst = self._alloc_reg()
        self._emit(EXPR_FCMP_EQ, dst, src1, src2)
        return dst

    def fcmp_ne(self, src1, src2):
        dst = self._alloc_reg()
        self._emit(EXPR_FCMP_NE, dst, src1, src2)
        return dst

    def fcmp_gt(self, src1, src2):
        dst = self._alloc_reg()
        self._emit(EXPR_FCMP_GT, dst, src1, src2)
        return dst

    def fcmp_ge(self, src1, src2):
        dst = self._alloc_reg()
        self._emit(EXPR_FCMP_GE, dst, src1, src2)
        return dst

    def fcmp_lt(self, src1, src2):
        dst = self._alloc_reg()
        self._emit(EXPR_FCMP_LT, dst, src1, src2)
        return dst

    def fcmp_le(self, src1, src2):
        dst = self._alloc_reg()
        self._emit(EXPR_FCMP_LE, dst, src1, src2)
        return dst

    # --- Boolean logic ---

    def bool_and(self, src1, src2):
        dst = self._alloc_reg()
        self._emit(EXPR_BOOL_AND, dst, src1, src2)
        return dst

    def bool_or(self, src1, src2):
        dst = self._alloc_reg()
        self._emit(EXPR_BOOL_OR, dst, src1, src2)
        return dst

    def bool_not(self, src):
        dst = self._alloc_reg()
        self._emit(EXPR_BOOL_NOT, dst, src, 0)
        return dst

    # --- Null checks ---

    def is_null(self, col_idx):
        dst = self._alloc_reg()
        self._emit(EXPR_IS_NULL, dst, col_idx, 0)
        return dst

    def is_not_null(self, col_idx):
        dst = self._alloc_reg()
        self._emit(EXPR_IS_NOT_NULL, dst, col_idx, 0)
        return dst

    # --- Type conversion ---

    def int_to_float(self, src):
        dst = self._alloc_reg()
        self._emit(EXPR_INT_TO_FLOAT, dst, src, 0)
        return dst

    # --- Output opcodes ---

    def emit_col(self, src_reg, payload_col_idx):
        self._emit(EXPR_EMIT, 0, src_reg, payload_col_idx)

    def copy_col(self, type_code, src_col_idx, payload_col_idx):
        self._emit(EXPR_COPY_COL, type_code, src_col_idx, payload_col_idx)

    def emit_null(self, payload_col_idx):
        self._emit(EXPR_EMIT_NULL, 0, payload_col_idx, 0)

    # --- String comparison ---

    def str_col_eq_const(self, col_idx, const_idx):
        dst = self._alloc_reg()
        self._emit(EXPR_STR_COL_EQ_CONST, dst, col_idx, const_idx)
        return dst

    def str_col_lt_const(self, col_idx, const_idx):
        dst = self._alloc_reg()
        self._emit(EXPR_STR_COL_LT_CONST, dst, col_idx, const_idx)
        return dst

    def str_col_le_const(self, col_idx, const_idx):
        dst = self._alloc_reg()
        self._emit(EXPR_STR_COL_LE_CONST, dst, col_idx, const_idx)
        return dst

    def str_col_eq_col(self, col_a, col_b):
        dst = self._alloc_reg()
        self._emit(EXPR_STR_COL_EQ_COL, dst, col_a, col_b)
        return dst

    def str_col_lt_col(self, col_a, col_b):
        dst = self._alloc_reg()
        self._emit(EXPR_STR_COL_LT_COL, dst, col_a, col_b)
        return dst

    def str_col_le_col(self, col_a, col_b):
        dst = self._alloc_reg()
        self._emit(EXPR_STR_COL_LE_COL, dst, col_a, col_b)
        return dst

    # --- Build ---

    def build(self, result_reg, const_strings=None):
        return ExprProgram(self._code, self._next_reg, result_reg,
                           const_strings=const_strings)

File: crates/gnitz-core/src/expr.rs
use gnitz_wire::{
    EXPR_BOOL_AND, EXPR_BOOL_NOT, EXPR_BOOL_OR, EXPR_CMP_EQ, EXPR_CMP_GE, EXPR_CMP_GT, EXPR_CMP_LE, EXPR_CMP_LT,
    EXPR_CMP_NE, EXPR_COPY_COL, EXPR_EMIT, EXPR_FCMP_EQ, EXPR_FCMP_GE, EXPR_FCMP_GT, EXPR_FCMP_LE, EXPR_FCMP_LT,
    EXPR_FCMP_NE, EXPR_FLOAT_ADD, EXPR_FLOAT_DIV, EXPR_FLOAT_MUL, EXPR_FLOAT_NEG, EXPR_FLOAT_SUB, EXPR_INT_ADD,
    EXPR_INT_DIV, EXPR_INT_MOD, EXPR_INT_MUL, EXPR_INT_NEG, EXPR_INT_SUB, EXPR_INT_TO_FLOAT, EXPR_IS_NOT_NULL,
    EXPR_IS_NULL, EXPR_LOAD_COL_FLOAT, EXPR_LOAD_COL_INT, EXPR_LOAD_CONST, EXPR_LOAD_NULL, EXPR_SELECT,
    EXPR_STR_COL_EQ_COL, EXPR_STR_COL_EQ_CONST, EXPR_STR_COL_LE_COL, EXPR_STR_COL_LE_CONST, EXPR_STR_COL_LT_COL,
    EXPR_STR_COL_LT_CONST,
};

/// A compiled expression program: a flat list of 4-word instructions
/// (opcode, dst_reg, arg1, arg2) plus metadata for embedding in filter params.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct ExprProgram {
    pub num_regs: u32,
    pub result_reg: u32,
    pub code: Vec<u32>,
    pub const_strings: Vec<String>,
}

impl ExprProgram {
    /// Serialise to a self-contained byte sequence (magic "EXPR"), suitable for a BLOB column.
    pub fn encode(&self) -> Vec<u8> {
        let strs: Vec<&[u8]> = self.const_strings.iter().map(String::as_bytes).collect();
        gnitz_wire::encode_expr_blob(self.num_regs, self.result_reg, &self.code, &strs)
    }
}

/// Builds an expression bytecode program with automatic register allocation.
pub struct ExprBuilder {
    code: Vec<u32>,
    next_reg: u32,
    const_strings: Vec<String>,
}

impl Default for ExprBuilder {
    fn default() -> Self {
        Self::new()
    }
}

impl ExprBuilder {
    pub fn new() -> Self {
        ExprBuilder {
            code: Vec::new(),
            next_reg: 0,
            const_strings: Vec::new(),
        }
    }

    fn alloc_reg(&mut self) -> u32 {
        let r = self.next_reg;
        self.next_reg += 1;
        r
    }

    fn emit(&mut self, op: u32, dst: u32, a1: u32, a2: u32) {
        self.code.extend_from_slice(&[op, dst, a1, a2]);
    }

    /// Allocate a destination register, emit a two-operand op into it, return it.
    fn binary_op(&mut self, op: u32, a: u32, b: u32) -> u32 {
        let dst = self.alloc_reg();
        self.emit(op, dst, a, b);
        dst
    }

    /// Allocate a destination register, emit a one-operand op into it, return it.
    /// A unary op is just a binary op whose second operand is unused.
    fn unary_op(&mut self, op: u32, a: u32) -> u32 {
        self.binary_op(op, a, 0)
    }

    pub fn load_col_int(&mut self, col_idx: usize) -> u32 {
        self.unary_op(EXPR_LOAD_COL_INT, col_idx as u32)
    }

    pub fn load_col_float(&mut self, col_idx: usize) -> u32 {
        self.unary_op(EXPR_LOAD_COL_FLOAT, col_idx as u32)
    }

    pub fn load_const(&mut self, value: i64) -> u32 {
        let (a1, a2) = gnitz_wire::encode_load_const(value);
        self.binary_op(EXPR_LOAD_CONST, a1, a2)
    }

    // --- Integer arithmetic ---

    pub fn add(&mut self, a: u32, b: u32) -> u32 {
        self.binary_op(EXPR_INT_ADD, a, b)
    }

    pub fn sub(&mut self, a: u32, b: u32) -> u32 {
        self.binary_op(EXPR_INT_SUB, a, b)
    }

    pub fn mul(&mut self, a: u32, b: u32) -> u32 {
        self.binary_op(EXPR_INT_MUL, a, b)
    }

    pub fn div(&mut self, a: u32, b: u32) -> u32 {
        self.binary_op(EXPR_INT_DIV, a, b)
    }

    pub fn modulo(&mut self, a: u32, b: u32) -> u32 {
        self.binary_op(EXPR_INT_MOD, a, b)
    }

    pub fn neg_int(&mut self, a: u32) -> u32 {
        self.unary_op(EXPR_INT_NEG, a)
    }

    // --- Float arithmetic ---

    pub fn float_add(&mut self, a: u32, b: u32) -> u32 {
        self.binary_op(EXPR_FLOAT_ADD, a, b)
    }

    pub fn float_sub(&mut self, a: u32, b: u32) -> u32 {
        self.binary_op(EXPR_FLOAT_SUB, a, b)
    }

    pub fn float_mul(&mut self, a: u32, b: u32) -> u32 {
        self.binary_op(EXPR_FLOAT_MUL, a, b)
    }

    pub fn float_div(&mut self, a: u32, b: u32) -> u32 {
        self.binary_op(EXPR_FLOAT_DIV, a, b)
    }

    pub fn float_neg(&mut self, a: u32) -> u32 {
        self.unary_op(EXPR_FLOAT_NEG, a)
    }

    // --- Integer comparison ---

    pub fn cmp_eq(&mut self, a: u32, b: u32) -> u32 {
        self.binary_op(EXPR_CMP_EQ, a, b)
    }

    pub fn cmp_ne(&mut self, a: u32, b: u32) -> u32 {
        self.binary_op(EXPR_CMP_NE, a, b)
    }

    pub fn cmp_gt(&mut self, a: u32, b: u32) -> u32 {
        self.binary_op(EXPR_CMP_GT, a, b)
    }

    pub fn cmp_ge(&mut self, a: u32, b: u32) -> u32 {
        self.binary_op(EXPR_CMP_GE, a, b)
    }

    pub fn cmp_lt(&mut self, a: u32, b: u32) -> u32 {
        self.binary_op(EXPR_CMP_LT, a, b)
    }

    pub fn cmp_le(&mut self, a: u32, b: u32) -> u32 {
        self.binary_op(EXPR_CMP_LE, a, b)
    }

    // --- Float comparison ---

    pub fn fcmp_eq(&mut self, a: u32, b: u32) -> u32 {
        self.binary_op(EXPR_FCMP_EQ, a, b)
    }

    pub fn fcmp_ne(&mut self, a: u32, b: u32) -> u32 {
        self.binary_op(EXPR_FCMP_NE, a, b)
    }

    pub fn fcmp_gt(&mut self, a: u32, b: u32) -> u32 {
        self.binary_op(EXPR_FCMP_GT, a, b)
    }

    pub fn fcmp_ge(&mut self, a: u32, b: u32) -> u32 {
        self.binary_op(EXPR_FCMP_GE, a, b)
    }

    pub fn fcmp_lt(&mut self, a: u32, b: u32) -> u32 {
        self.binary_op(EXPR_FCMP_LT, a, b)
    }

    pub fn fcmp_le(&mut self, a: u32, b: u32) -> u32 {
        self.binary_op(EXPR_FCMP_LE, a, b)
    }

    // --- Boolean ---

    pub fn bool_and(&mut self, a: u32, b: u32) -> u32 {
        self.binary_op(EXPR_BOOL_AND, a, b)
    }

    pub fn bool_or(&mut self, a: u32, b: u32) -> u32 {
        self.binary_op(EXPR_BOOL_OR, a, b)
    }

    pub fn bool_not(&mut self, a: u32) -> u32 {
        self.unary_op(EXPR_BOOL_NOT, a)
    }

    // --- Null checks ---

    pub fn is_null(&mut self, col_idx: usize) -> u32 {
        self.unary_op(EXPR_IS_NULL, col_idx as u32)
    }

    pub fn is_not_null(&mut self, col_idx: usize) -> u32 {
        self.unary_op(EXPR_IS_NOT_NULL, col_idx as u32)
    }

    // --- Type conversion ---

    pub fn int_to_float(&mut self, src: u32) -> u32 {
        self.unary_op(EXPR_INT_TO_FLOAT, src)
    }

    // --- Conditional ---

    pub fn select(&mut self, cond: u32, a: u32, b: u32) -> u32 {
        let dst = self.alloc_reg();
        self.emit(EXPR_SELECT, dst, cond, gnitz_wire::encode_select_operands(a, b));
        dst
    }

    /// Materialize a NULL value into a fresh register (value 0, null bit set).
    pub fn load_null(&mut self) -> u32 {
        let dst = self.alloc_reg();
        self.emit(EXPR_LOAD_NULL, dst, 0, 0);
        dst
    }

    // --- Output opcodes ---

    pub fn emit_col(&mut self, src_reg: u32, payload_col_idx: u32) {
        self.emit(EXPR_EMIT, 0, src_reg, payload_col_idx);
    }

    pub fn copy_col(&mut self, type_code: u32, src_col_idx: u32, payload_col_idx: u32) {
        self.emit(EXPR_COPY_COL, type_code, src_col_idx, payload_col_idx);
    }

    // --- String constants ---

    pub fn add_const_string(&mut self, s: String) -> u32 {
        let idx = self.const_strings.len() as u32;
        self.const_strings.push(s);
        idx
    }

    // --- String comparisons ---

    pub fn str_col_eq_const(&mut self, col_idx: usize, const_idx: u32) -> u32 {
        self.binary_op(EXPR_STR_COL_EQ_CONST, col_idx as u32, const_idx)
    }

    pub fn str_col_lt_const(&mut self, col_idx: usize, const_idx: u32) -> u32 {
        self.binary_op(EXPR_STR_COL_LT_CONST, col_idx as u32, const_idx)
    }

    pub fn str_col_le_const(&mut self, col_idx: usize, const_idx: u32) -> u32 {
        self.binary_op(EXPR_STR_COL_LE_CONST, col_idx as u32, const_idx)
    }

    pub fn str_col_eq_col(&mut self, col_a: usize, col_b: usize) -> u32 {
        self.binary_op(EXPR_STR_COL_EQ_COL, col_a as u32, col_b as u32)
    }

    pub fn str_col_lt_col(&mut self, col_a: usize, col_b: usize) -> u32 {
        self.binary_op(EXPR_STR_COL_LT_COL, col_a as u32, col_b as u32)
    }

    pub fn str_col_le_col(&mut self, col_a: usize, col_b: usize) -> u32 {
        self.binary_op(EXPR_STR_COL_LE_COL, col_a as u32, col_b as u32)
    }

    pub fn build(self, result_reg: u32) -> ExprProgram {
        ExprProgram {
            num_regs: self.next_reg,
            result_reg,
            code: self.code,
            const_strings: self.const_strings,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn encode_round_trips_through_wire_decoder() {
        let mut b = ExprBuilder::new();
        let c = b.load_const(1_234_567_890_123);
        let col = b.load_col_int(0);
        let cond = b.cmp_gt(col, c);
        let s_idx = b.add_const_string("längre sträng".to_string());
        let _ = b.str_col_eq_const(1, s_idx);
        let _ = b.add_const_string(String::new());
        let sel = b.select(cond, col, c);
        let prog = b.build(sel);

        let blob = prog.encode();
        let dec = gnitz_wire::decode_expr_blob(&blob).unwrap();
        assert_eq!(dec.num_regs, prog.num_regs);
        assert_eq!(dec.result_reg, prog.result_reg);
        assert_eq!(dec.code, prog.code);
        let expected: Vec<Vec<u8>> = prog.const_strings.iter().map(|s| s.as_bytes().to_vec()).collect();
        assert_eq!(dec.const_strings, expected);
    }
}
